src/runtime_src/core/common/xrt_xclbin.cpp

```diff
@@ -74,3 +79,375 @@ xrtXclbinUUID(xclDeviceHandle dhdl, xuid_t out)
     return -1;
   }
 }
+
+namespace xrt {
+
+// class xclbin_impl - class for xclbin objects
+//
+// Life time of xclbin are managed through shared pointers.
+// A buffer is freed when last references is released.
+class xclbin_impl
+{
+protected:
+  std::vector<char> m_axlf;
+  const axlf* m_top;
+
+  void init_axlf_handle()
+  {
+    const axlf* tmp = reinterpret_cast<const axlf*>(m_axlf.data());
+    if (strncmp(tmp->m_magic, "xclbin2", 7)) // Future: Do not hardcode "xclbin2"
+      throw std::runtime_error("Invalid xclbin");
+    m_top = tmp;
+  }
+
+public:
+  xclbin_impl() : m_top(nullptr) {}
+
+  xclbin_impl(const std::vector<char>& data)
+  {
+    m_axlf = data;
+    // Set pointer of type axlf*, and check magic string
+    init_axlf_handle();
+  }
+
+  xclbin_impl(const std::string& filename)
+  {
+    if (filename.empty())
+      throw std::runtime_error("No XCLBIN specified");
+
+    // Load the file into data
+    std::ifstream stream(filename);
+    stream.seekg(0, stream.end);
+    size_t size = stream.tellg();
+    stream.seekg(0, stream.beg);
+    m_axlf.resize(size);
+    stream.read(m_axlf.data(), size);
+
+    // Set pointer of type axlf*, and check magic string
+    init_axlf_handle();
+  }
+
+  bool
+  empty() const
+  {
+    return m_axlf.size() == 0 || m_top == nullptr;
+  }
+
+  void
+  not_empty_or_error() const
+  {
+    if (empty())
+      throw xrt_core::error(-EINVAL, "Invalid XCLBIN data");
+  }
+
+  std::vector<std::string>
+  get_cu_names() const
+  {
+    std::vector<std::string> names;
+    for (auto cu : xrt_core::xclbin::get_cus(m_top, false)) // Why would I ever need them encoded
+      names.push_back(xrt_core::xclbin::get_ip_name(m_top, cu));
+
+    return names;
+  }
+
+  std::string
+  get_xsa_name() const
+  {
+    not_empty_or_error();
+    return reinterpret_cast<const char*>(m_top->m_header.m_platformVBNV);
+  }
+
+  uuid
+  get_uuid() const
+  {
+    not_empty_or_error();
+    return {m_top->m_header.uuid};
+  }
+
+  const std::vector<char>&
+  get_data() const
+  {
+    not_empty_or_error();
+    return m_axlf;
+  }
+};
+
+} //namespace
+
+////////////////////////////////////////////////////////////////
+// xrt_xclbin C++ API implementations (xrt_xclbin.h)
+////////////////////////////////////////////////////////////////
+namespace xrt {
+
+xclbin::
+xclbin()
+  : handle(std::make_shared<xclbin_impl>())
+{}
```

Comment:
I think the default constructor of xclbin should not construct an xclbin_impl, it should behave as a nullptr, e.g. xrt::xclbin::handle member should be a nullptr.   An xclbin_impl should always be valid.

If the user does somethng like this:
```
xrt::xclbin x;
x.get_cu_names();
```
then it is a user error.